x-pack/plugin/security/qa/jwt-realm/src/javaRestTest/java/org/elasticsearch/xpack/security/authc/jwt/JwtRestIT.java

```diff
@@ -0,0 +1,458 @@
+        return signJWt(signer, "RS256", claimsSet);
+    }
+
+    private RSASSASigner loadRsaSigner() throws IOException, ParseException, JOSEException {
+        // The "jwt1" realm is configured using public JWKSet (in build.gradle)
+        try (var in = getDataInputStream("/jwk/rsa-private-jwkset.json")) {
+            final JWKSet jwkSet = JWKSet.load(in);
+            final JWK key = jwkSet.getKeyByKeyId("test-rsa-key");
+            assertThat(key, instanceOf(RSAKey.class));
+            return new RSASSASigner((RSAKey) key);
+        }
+    }
+
+    private SignedJWT signHmacJwt(JWTClaimsSet claimsSet) throws JOSEException, ParseException {
+        // Input string is configured in build.gradle
+        return signHmacJwt(claimsSet, "test-HMAC/secret passphrase-value");
+    }
+
+    private SignedJWT signHmacJwt(JWTClaimsSet claimsSet, String hmacPassphrase) throws JOSEException, ParseException {
+        final OctetSequenceKey hmac = JwkValidateUtil.buildHmacKeyFromString(hmacPassphrase);
+        final JWSSigner signer = new MACSigner(hmac);
+        return signJWt(signer, "HS256", claimsSet);
+    }
+
+    // JWT construction
+    private JWTClaimsSet buildJwt(Map<String, Object> claims, Instant issueTime) {
+        final JWTClaimsSet.Builder builder = new JWTClaimsSet.Builder();
+        builder.issuer(randomAlphaOfLengthBetween(4, 24));
+        builder.subject(randomAlphaOfLengthBetween(4, 24));
+        builder.audience(randomList(1, 6, () -> randomAlphaOfLengthBetween(4, 12)));
+        if (randomBoolean()) {
+            builder.jwtID(UUIDs.randomBase64UUID(random()));
+        }
+
+        issueTime = issueTime.truncatedTo(ChronoUnit.SECONDS);
+        builder.issueTime(Date.from(issueTime));
+        builder.expirationTime(Date.from(issueTime.plusSeconds(randomLongBetween(180, 1800))));
+        if (randomBoolean()) {
+            builder.notBeforeTime(Date.from(issueTime.minusSeconds(randomLongBetween(10, 90))));
+        }
+
+        // This may overwrite the aud/sub/iss set above. That is an intended behaviour
+        for (String key : claims.keySet()) {
+            builder.claim(key, claims.get(key));
+        }
+        return builder.build();
+    }
+
+    private SignedJWT signJWt(JWSSigner signer, String algorithm, JWTClaimsSet claimsSet) throws JOSEException {
+        final JWSHeader jwtHeader = new JWSHeader.Builder(JWSAlgorithm.parse(algorithm)).build();
+        final SignedJWT jwt = new SignedJWT(jwtHeader, claimsSet);
+        jwt.sign(signer);
+        return jwt;
+    }
+
+    private TestSecurityClient getSecurityClient(SignedJWT jwt, Optional<String> sharedSecret) {
+        final String bearerHeader = "Bearer " + jwt.serialize();
+
+        final RequestOptions.Builder options = RequestOptions.DEFAULT.toBuilder().addHeader("Authorization", bearerHeader);
+        sharedSecret.ifPresent(secret -> options.addHeader("X-Client-Authentication", "SharedSecret " + secret));
```

Comment:
Nit: Use JwtRealm.HEADER_CLIENT_AUTHENTICATION. This header value is subject to change, and using the constant will avoid having to apply a change to multiple classes.

Nit: Use JwtRealmSettings.CLIENT_AUTHENTICATION_TYPE_SHARED_SECRET. I think you proposed renaming the value in a different PR. Using the constant will avoid having to apply the change to multiple classes.